### PR TITLE
IR-based inliner

### PR DESCRIPTION
**Pull request by gramalingam**
Add utility to inline all functions in a model.

Still TODO:

* Some edge cases to be considered for renaming and avoiding conflicts, especially with subgraphs.
* Must ensure no variable capture happens (part of above renaming).
* Test renaming of node names.

Fixes https://github.com/microsoft/onnxscript/issues/1769

### CODE REVIEW

File: onnxscript/optimizer/_inliner.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+            doc_string=copied.doc_string,
+            metadata_props=copied.metadata_props,
+        )
+        new_outputs = new_node.outputs
+        for i, output in enumerate(copied.outputs):
+            self._value_map[output] = new_outputs[i]
+            new_outputs[i].name = output.name
+        return new_node
+
+    def clone_graph(self, graph: ir.Graph) -> ir.Graph:
```
Comment (justinchuby) [inline onnxscript/optimizer/_inliner.py:78]: Thanks! I wonder if we should implement a `copy()` method in Graph, or implement the `__deepcopy__` magic for it to provide pythonic apis?

Comment (gramalingam) [reply onnxscript/optimizer/_inliner.py:78]: It is not similar to standard copy, though. It is dependent on some state/parameters (mappings for outer-scope references inside the graph/node).

**Comment by justinchuby** [comment]
Just checking: will the meta data from the function call node be added to the inlined nodes? That would be desirable.

**Comment by justinchuby** [comment]
In the torch exporter's case, since only nodes from the main graph have metadata written to them, taking the metadata from the main node is the most useful.

**Comment by gramalingam** [comment]
Added merging of metadata props

File: onnxscript/optimizer/_inliner.py [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+            if id in self._functions:
+                # If there are multiple calls to same function, we use a prefix to disambiguate
+                # the different call-sites:
+                if id_count[id] > 1:
+                    call_site_prefix = f"_{next_id[id]}"
+                    next_id[id] += 1
+                else:
+                    call_site_prefix = ""
+                call_site = node.name or (self._function_id_abbreviations[id] + call_site_prefix)
+                nodes, values = self._instantiate_call(node, call_site)
```
Comment (justinchuby) [inline onnxscript/optimizer/_inliner.py:252]: Does the CallSiteId need to follow specific formats? Would it be a good idea to create a function that constructs it?

File: onnxscript/optimizer/_inliner.py [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+    return {id: id_abbreviation(id) for id in function_ids}
+
+class _Inliner:
+    def __init__(self, model: ir.Model) -> None:
+        self._functions = model.functions
+        self._function_id_abbreviations = _abbreviate(self._functions.keys())
+        self._opset_imports = model.opset_imports
+        self.used_value_names: set[str] = set()
+        self.used_node_names: set[str] = set()
+        self.node_context: dict[ir.Node, CallStack] = {}
```
Comment (justinchuby) [inline onnxscript/optimizer/_inliner.py:171]: Can they be made private?

File: onnxscript/optimizer/_inliner.py [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+                if id_count[id] > 1:
+                    call_site_prefix = f"_{next_id[id]}"
+                    next_id[id] += 1
+                else:
+                    call_site_prefix = ""
+                call_site = node.name or (self._function_id_abbreviations[id] + call_site_prefix)
+                nodes, values = self._instantiate_call(node, call_site)
+                convenience.replace_nodes_and_values(
+                    graph, node, [node], nodes, node.outputs, values
+                )
```
Comment (justinchuby) [inline onnxscript/optimizer/_inliner.py:255]: would be good to specify keyword arguments for readability

File: onnxscript/optimizer/_inliner.py [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+            else:
```
Comment (justinchuby) [inline onnxscript/optimizer/_inliner.py:256]: Does this assume a function call will not take graph attributes?

File: onnxscript/optimizer/_inliner.py (hunk 1 above)
Comment (gramalingam) [reply onnxscript/optimizer/_inliner.py:133]: What is done here is not suitable for copy/clone and vice-versa.  Different logic. At best, we could rename these methods and call them by another name like "apply_substitutions" or "apply_replacements". I thought that the name of the class "CopyReplace" would suffice to indicate what's meant, but it can be improved.

Comment (justinchuby) [reply onnxscript/optimizer/_inliner.py:133]: I see. Thanks!

File: onnxscript/optimizer/_inliner.py [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+            new_outputs[i].name = _make_unique_name(old_name, self._call_stack, self._inliner.used_value_names)
+
+        self._inliner.node_context[new_node] = self._call_stack
+
+        return new_node
+
+    def clone_graph(self, graph: ir.Graph) -> ir.Graph:
+        input_values = [self.clone_value(v) for v in graph.inputs]
+        nodes = [self.clone_node(node) for node in graph]
+        initializers = graph.initializers  # Initializers are not cloned, but shared.
```
[the remark this answers is not in the document]
Comment (gramalingam) [reply onnxscript/optimizer/_inliner.py:136]: It's happening already a few lines below, this is dead code ... I will remove it, thanks.

File: onnxscript/optimizer/_inliner.py (hunk 3 above)
Comment (gramalingam) [reply onnxscript/optimizer/_inliner.py:171]: They are used in the auxiliary class _CopyReplace above. I made this class (Inliner) itself private, so hopefully it is clear that this is meant to be internal to the file (but usable by other classes in the file). Not sure if there is a better convention for this.

Comment (justinchuby) [reply onnxscript/optimizer/_inliner.py:171]: I see - I didn't realized they are used outside. Thanks!

File: onnxscript/optimizer/_inliner.py (hunk 5 above)
Comment (gramalingam) [reply onnxscript/optimizer/_inliner.py:256]: Good question. I don't think this is relevant to the code here in the else branch, which is not inside a function. But the question is relevant to what happens in the code in CopyReplace, when we do substitution of attribute values for attribute parameters. I think that code does assume that a function call will not take graph attributes for now ... I think it will cause an assertion failure (in some cases). Generalizing to handle that would be interesting ... doable, I think, but the complexity is mostly in ensuring correctness.

File: onnxscript/optimizer/_inliner.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (gramalingam) [reply onnxscript/optimizer/_inliner.py:136]: Maybe not ... I didn't understand what you meant by const_value, but I guess you are suggesting another layer of copying, will do

File: onnxscript/optimizer/_inliner.py [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+            return attr
+        assert isinstance(attr, ir.RefAttr)
+        if key in self._attr_map:
+            return self._attr_map[key]
+        # Note that if a function has an attribute-parameter X, and a call (node) to the function
+        # has no attribute X, all references to X in nodes inside the function body will be
+        # removed. This is just the ONNX representation of optional-attributes.
+        return None
+
+    def clone_node(self, copied: ir.Node) -> ir.Node:
```
[the remark this answers is not in the document]
Comment (gramalingam) [reply onnxscript/optimizer/_inliner.py]: done

File: onnxscript/optimizer/_inliner.py [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+        self._attr_map = attr_map
+        self._metadata_props = metadata_props
+        self._call_stack = call_stack
+
+    def clone_value(self, value: ir.Value) -> ir.Value | None:
+        if value in self._value_map:
+            return self._value_map[value]
+        # If the value is not in the value map, it must be a graph input.
+        assert value.producer() is not None, f"Value {value} has no entry in the value map"
+        new_value = ir.Value(
```
[the remark this answers is not in the document]
Comment (gramalingam) [reply onnxscript/optimizer/_inliner.py:66]: done

File: onnxscript/optimizer/_inliner.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (gramalingam) [reply onnxscript/optimizer/_inliner.py]: done

File: onnxscript/optimizer/_inliner.py (hunk 4 above)
Comment (gramalingam) [reply onnxscript/optimizer/_inliner.py]: done

File: onnxscript/optimizer/_inliner.py [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+        # Do we need to preserve both?
+
+        new_node = ir.Node(
+            copied.domain,
+            copied.op_type,
+            new_inputs,
+            new_attributes,
+            overload=copied.overload,
+            num_outputs=len(copied.outputs),
+            graph=None,  #  TODO:
```
[the remark this answers is not in the document]
Comment (gramalingam) [reply onnxscript/optimizer/_inliner.py:118]: That was a reminder to check if we need to do anything. I guess this should be fine, right? Will this be set for nodes when we construct a new `Graph(nodes=nodes, ...)` ? And, similarly, when we insert a new node into a graph?

File: onnxscript/optimizer/_inliner.py [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+"""Implementation of an inliner for onnxscript.ir"""
+
+from __future__ import annotations
+
+from collections import defaultdict
+from dataclasses import dataclass, field
+from typing import Iterable, Sequence, Tuple
+
+import onnxscript.ir as ir
+import onnxscript.ir.convenience as convenience
```
[the remark this answers is not in the document]
Comment (gramalingam) [reply onnxscript/optimizer/_inliner.py:12]: Wouldn't `ir.convenience` be better?

File: onnxscript/optimizer/_inliner.py (hunk 2 above)
Comment (gramalingam) [reply onnxscript/optimizer/_inliner.py:252]: If you mean does the inliner have any requirements, no. But part of my earlier question about the modularization logic was whether there were any conventions established (which might be useful across various tools if we stick to one) for generating qualified names. (It is a bit tricky, though, to get the ordering right ... when we start composing names like "layer_4_activation = f (layer_4_matmul_output)` and we inline `f`.)

File: onnxscript/optimizer/_inliner.py [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+                    if attr.type == ir.AttributeType.GRAPH:
+                        self.inline_calls_in(attr.value)
+                    elif attr.type == ir.AttributeType.GRAPHS:
+                        for graph in attr.value:
+                            self.inline_calls_in(graph)
+
+
+
+def inline(model: ir.Model) -> None:
+    """Inline all function calls (recursively) in the model."""
```
[the remark this answers is not in the document]
Comment (gramalingam) [reply onnxscript/optimizer/_inliner.py:269]: Should be doable (if and when we want to do it; not sure it's worthwhile right now).

File: onnxscript/optimizer/_inliner.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (justinchuby) [reply onnxscript/optimizer/_inliner.py:118]: Yes - that’s correct for both cases

File: onnxscript/optimizer/_inliner.py [hunk 12]
```diff
@@ -0,0 +1,272 @@
...
+    def clone_graph(self, graph: ir.Graph) -> ir.Graph:
+        input_values = [self.clone_value(v) for v in graph.inputs]
+        nodes = [self.clone_node(node) for node in graph]
+        initializers = graph.initializers  # Initializers are not cloned, but shared.
+
+        return ir.Graph(
+            input_values,  # type: ignore
+            graph.outputs,
+            nodes=nodes,
+            initializers=list(initializers.values()),
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply onnxscript/optimizer/_inliner.py]: Sorry I was wrong I think.

File: onnxscript/optimizer/_inliner.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (justinchuby) [reply onnxscript/optimizer/_inliner.py]: I might be wrong. Will come back tomorrow. Thanks!

File: onnxscript/optimizer/_inliner.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (justinchuby) [reply onnxscript/optimizer/_inliner.py:12]: Yes - we need to expose that

File: onnxscript/optimizer/_inliner.py [hunk 13]
```diff
@@ -0,0 +1,272 @@
...
+
+        return new_node
+
+    def clone_graph(self, graph: ir.Graph) -> ir.Graph:
+        input_values = [self.clone_value(v) for v in graph.inputs]
+        nodes = [self.clone_node(node) for node in graph]
+        initializers = graph.initializers  # Initializers are not cloned, but shared.
+
+        return ir.Graph(
+            input_values,  # type: ignore
```
[the remark this answers is not in the document]
Comment (gramalingam) [reply onnxscript/optimizer/_inliner.py:139]: No ... the problem is with clone_value ... which returns None, but it won't here.

File: onnxscript/optimizer/_inliner.py [hunk 14]
```diff
@@ -0,0 +1,272 @@
...
+            nodes=nodes,
+            initializers=list(initializers.values()),
+            doc_string=graph.doc_string,
+            opset_imports=graph.opset_imports,
+            name=graph.name,
+            metadata_props=graph.metadata_props,
+        )
+
+def _abbreviate(function_ids: Iterable[ir.OperatorIdentifier]) -> dict[ir.OperatorIdentifier, str]:
+    """Create a short unambiguous abbreviation for all function ids."""
```
[the remark this answers is not in the document]
Comment (gramalingam) [reply onnxscript/optimizer/_inliner.py:150]: Will skip for now. A complete description will be as complex as the code below. And it is not clear that we want to stick to that spec, we may want to change it too. Eg.: should overload be suppressed, if it is unambiguous? May be, may be not.

**Review by justinchuby — APPROVED** [review]
Maybe functions cannot have initialized values, so const_values are irrelevant after all?
